### ISSUE TITLE
'client_secret' parameter is not valid (Cash Bounty for fixer!)

### ISSUE THREAD

**Issue by AmitMY** [lelylan/simple-oauth2#185]
Using `simple-oauth2@1.5.2` I am trying to get a token after getting oAuth code from bing, like:

```ts
    this.client = oauth2.create({
      client: {
        id: conf.client_id,
        secret: conf.client_secret
      },
      auth: {
        tokenHost: conf.token_host,
        tokenPath: conf.token_path,
        authorizePath: conf.authorize_path
      }
    });
    const result = await this.client.authorizationCode.getToken({ code: req.query.code });
```

And every time I get this error: 
> The provided value for the 'client_secret' parameter is not valid.

I have made sure the parameter is correct and even generated a new one. Is there something trick I need to do? Like, encode some value?

* I can't upgrade to v2, because I need an es5 distribution and you don't support it yet (`async` and `await` don't compile for firebase cloud functions if it is the `node_modules`)

**Comment by jonathansamines** [comment]
Do you mind to provide an example of what kind of value are you trying to provide as "secret". A reproducible example would be ideal.

**Comment by AmitMY** [comment]
Generated a test app:
```
id = '59a9e5ed-f20a-4130-a97c-9b537671ac4e'
secret = 'qcFD88!maxuzABKOR173(*]'
```

I can send repro code, for a firebase function that uses simple-oauth

**Comment by AmitMY** [comment]
To create an app go to: https://apps.dev.microsoft.com/
It's really simple, but simple-oauth2 doesn't work for it

**Comment by jonathansamines** [comment]
@AmitMY I just tried your test configuration, and even tried with a new client secret I just generated from the microsoft dev center and wasn't able to reproduce your issue using `simple-oauth2@1.5.2`. Please provide a reproduction repository so that I can better help you.

**Comment by AmitMY** [comment]
Reproduction:
```js
const oauth2 = require('simple-oauth2'); // npm i simple-oauth2@1.5.2

id = "" // ID goes here
secret = "" // Secret goes here

const client = oauth2.create({
  client: {
	id: id,
	secret: secret
  },
  auth: {
	tokenHost: "https://login.live.com/",
	tokenPath: "oauth20_token.srf",
	authorizePath: "oauth20_authorize.srf"
  }
});

client.authorizationCode.getToken({ code: "nothing" })
	.then(res => console.log(res))
	.catch(e => console.error(e))
```

Run with node:
> { Error: Unauthorized
  name: 'Error',
  status: 401,
  message: 'Unauthorized',
  context:
   { error: 'invalid_client',
     error_description: 'The provided value for the \'client_secret\' parameter is not valid.' } }

**Comment by DecentGradient** [comment]
client - required object with the following properties:

id - Service registered client id. Required.
secret - Service registered client secret. Required.
**secretParamName - Parameter name used to send the client secret. Default to client_secret.
idParamName - Parameter name used to send the client id. Default to client_id.**

**Comment by AmitMY** [comment]
@DecentGradient As you can see, it complains about the "client_secret" field which is also the default field value for `secretParamName`.

I would not investigate more in this direction, because using `node-oauth` works defaultly.

**Comment by jonathansamines** [comment]
@AmitMY I finally got some time to test this out. I was able to properly use the microsoft authorization server, however it required some additional configuration to get it working:

- Microsoft requires the authorization credentials for the token exchange to be sent in the body. simple-oauth@1.x.x allows you to configure `options.useBodyAuth`. See the [docs](https://github.com/lelylan/simple-oauth2/tree/1.x#options)
- Microsoft requires to send the `redirect_uri` parameter during the token exchange. 

With those additional, I was able to get a valid token. I put together a [microsoft example](https://github.com/lelylan/simple-oauth2/blob/develop/example/microsoft.js), however I am using simple-oauth2@2.x.x because I am no longer putting work on 1.x.x unless a bug is found. The configuration should be pretty similar though.
